<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-исследования" data-toc-modified-id="Описание-исследования-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание исследования</a></span></li><li><span><a href="#Цель-исследования" data-toc-modified-id="Цель-исследования-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Цель исследования</a></span></li><li><span><a href="#Задачи-исследования" data-toc-modified-id="Задачи-исследования-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Задачи исследования</a></span></li><li><span><a href="#Исходные-данные" data-toc-modified-id="Исходные-данные-0.4"><span class="toc-item-num">0.4&nbsp;&nbsp;</span>Исходные данные</a></span></li><li><span><a href="#Стейкхолдеры" data-toc-modified-id="Стейкхолдеры-0.5"><span class="toc-item-num">0.5&nbsp;&nbsp;</span>Стейкхолдеры</a></span></li></ul></li><li><span><a href="#Изучение-данных-из-файла-" data-toc-modified-id="Изучение-данных-из-файла--1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение данных из файла <a id="1-bullet" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Визуализация-данных" data-toc-modified-id="Визуализация-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Визуализация данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Кластеризация-клиентов" data-toc-modified-id="Кластеризация-клиентов-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Кластеризация клиентов</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Проверка-гипотез" data-toc-modified-id="Проверка-гипотез-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка гипотез</a></span></li></ul></li><li><span><a href="#Выводы-и-рекомендации" data-toc-modified-id="Выводы-и-рекомендации-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы и рекомендации</a></span></li><li><span><a href="#Дашборд-и-презентация" data-toc-modified-id="Дашборд-и-презентация-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Дашборд и презентация</a></span></li></ul></div>

# Описание проекта

### Описание исследования
Предоставлены данные о клиентах банка «Метанпром». Банк располагается в Ярославле и областных городах: Ростов Великий и Рыбинск. В исследовании будут проанализированы клиенты регионального банка и выделен портрет клиентов, которые склонны уходить из банка.
    
_____
### Цель исследования
Дать рекомендации по повышению лояльности клиентов банка и удержанию их в числе постоянных клиентов на основе аналитических данных и материалов по проверке гипотез. 
    
_____
### Задачи исследования

1. Провести предварительную подготовку данных:
- исследовать пропущенные значения;
- исследовать соответствие типов;
- исследовать дубликаты;
- проверить корректность наименований колонок;
- переименовать колонки, если необходимо;
- удалить дубликаты;
- привести типы;
- заменить пропущенные значения.
2. Проанализировать полученные данные. 
3. Проверить гипотезы:
* >Гипотеза различия среднего дохода между экспериментальными группами клиентов банка по числу используемых услуг:
* Н0 - различия среднего дохода между экспериментальными группами клиентов банка которые пользуются двумя и более продуктами банка, и теми, которые пользуются одним, статистически не значимы
* Н1 - различия среднего дохода между экспериментальными группами клиентов банка которые пользуются двумя и более продуктами банка, и теми, которые пользуются одним, статистически, статистически значимы
* >Гипотеза относительно представленных данных.

Для формирования гипотез следует решить подзадачи:  
3.1. Рассмотреть распределение и важность признаков в группах клиентов (ушедших и оставшихся);  
3.2. Определить зависимость частоты и числа покупки услуг(банковских продуктов) и вероятности ухода;  
3.3. Сегментировать группы клиентов по объему потребляемых услуг и составить портреты клиентов для разработки рекомендаций по удержанию;  
3.4. Проверить гипотезы и разработать конкретные рекомендации по удержанию клиентских групп.  
4. Подготовить дашборд и презентацию.
_____
### Исходные данные
    
Данные клиентов банка:

- `userid` — идентификатор пользователя,
- `score` — баллы кредитного скоринга,
- `City` — город,
- `Gender` — пол,
- `Age` — возраст,
- `Objects` — количество объектов в собственности,
- `Balance` — баланс на счёте,
- `Products` — количество продуктов, которыми пользуется клиент,
- `CreditCard` — есть ли кредитная карта,
- `Loyalty` — активный клиент,
- `estimated_salary` — заработная плата клиента,
- `Churn` — ушёл или нет.
______
### Стейкхолдеры
    
`Заказчик и основной пользователь` - менеджер продукта, разрабатывающий решения по удержанию клиентов.


## Изучение данных из файла <a id='1-bullet'></a>

In [1]:
!pip install seaborn --upgrade  
!pip install seaborn

  Attempting uninstall: seaborn
    Found existing installation: seaborn 0.11.2
    Uninstalling seaborn-0.11.2:
      Successfully uninstalled seaborn-0.11.2


In [2]:
#импортируем бибилиотеки
import pandas as pd
import seaborn as sns

from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso, Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

import matplotlib.pyplot as plt
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

import numpy as np
from IPython.display import display

from scipy import stats as st
import numpy as np
import math as mth
import plotly.express as px

from statsmodels.stats.proportion import proportions_ztest
import warnings
warnings.filterwarnings('ignore')

In [3]:
try:
    bank_data = pd.read_csv('bank_dataset.csv', sep=',')
except:
    bank_data = pd.read_csv('/datasets/bank_dataset.csv', sep=',')

display(bank_data.head())
#сохраним данные для дашборда
bank_data.to_csv('final_tableau.csv', index = False)

userid  score       City Gender  Age  Objects    Balance  Products  \
0  15677338    619  Ярославль      Ж   42        2        NaN         1   
1  15690047    608    Рыбинск      Ж   41        1   83807.86         1   
2  15662040    502  Ярославль      Ж   42        8  159660.80         3   
3  15744090    699  Ярославль      Ж   39        1        NaN         2   
4  15780624    850    Рыбинск      Ж   43        2  125510.82         1   

   CreditCard  Loyalty  estimated_salary  Churn  
0           1        1         101348.88      1  
1           0        1         112542.58      0  
2           1        0         113931.57      1  
3           0        0          93826.63      0  
4           1        1          79084.10      0

In [4]:
#изменим название колонок
bank_data = bank_data.rename(columns= {"CreditCard": "сredit_сard"})

#общая функция для предобработки, дат в нашем датасете нет, но я не удаляю эту часть
def inspect(df):
    display(df.head()) 
    df.info() 
    print ('Количество дубликатов -', df.duplicated().sum())
    numeric_columns=df.describe().columns
    display(df.describe())
    display()
    df.columns = [x.lower() for x in df.columns]
    for column_name in df.columns:
        if column_name not in numeric_columns:
            if 'session' in column_name or 'dt' in column_name:
                df[column_name] = pd.to_datetime(df[column_name])
            else:
                display(df[column_name].value_counts())
                print(20*'*')
    return df

bank_data.columns = inspect(bank_data)



userid  score       City Gender  Age  Objects    Balance  Products  \
0  15677338    619  Ярославль      Ж   42        2        NaN         1   
1  15690047    608    Рыбинск      Ж   41        1   83807.86         1   
2  15662040    502  Ярославль      Ж   42        8  159660.80         3   
3  15744090    699  Ярославль      Ж   39        1        NaN         2   
4  15780624    850    Рыбинск      Ж   43        2  125510.82         1   

   сredit_сard  Loyalty  estimated_salary  Churn  
0            1        1         101348.88      1  
1            0        1         112542.58      0  
2            1        0         113931.57      1  
3            0        0          93826.63      0  
4            1        1          79084.10      0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
userid              10000 non-null int64
score               10000 non-null int64
City                10000 non-null object
Gender              10000 non-null object
Age                 10000 non-null int64
Objects             10000 non-null int64
Balance             6383 non-null float64
Products            10000 non-null int64
сredit_сard         10000 non-null int64
Loyalty             10000 non-null int64
estimated_salary    10000 non-null float64
Churn               10000 non-null int64
dtypes: float64(2), int64(8), object(2)
memory usage: 937.6+ KB
Количество дубликатов - 0


userid         score           Age       Objects        Balance  \
count  1.000000e+04  10000.000000  10000.000000  10000.000000    6383.000000   
mean   1.573368e+07    650.528800     38.921800      5.012800  119827.493793   
std    7.193619e+04     96.653299     10.487806      2.892174   30095.056462   
min    1.560844e+07    350.000000     18.000000      0.000000    3768.690000   
25%    1.567126e+07    584.000000     32.000000      3.000000  100181.975000   
50%    1.573347e+07    652.000000     37.000000      5.000000  119839.690000   
75%    1.579597e+07    718.000000     44.000000      7.000000  139512.290000   
max    1.585843e+07    850.000000     92.000000     10.000000  250898.090000   

           Products  сredit_сard       Loyalty  estimated_salary         Churn  
count  10000.000000  10000.00000  10000.000000      10000.000000  10000.000000  
mean       1.530200      0.70550      0.515100     100090.239881      0.203700  
std        0.581654      0.45584      0.499797      57510.492818      0.402769  
min        1.000000      0.00000      0.000000         11.580000      0.000000  
25%        1.000000      0.00000      0.000000      51002.110000      0.000000  
50%        1.000000      1.00000      1.000000     100193.915000      0.000000  
75%        2.000000      1.00000      1.000000     149388.247500      0.000000  
max        4.000000      1.00000      1.000000     199992.480000      1.000000

Ярославль         5014
Ростов Великий    2509
Рыбинск           2477
Name: city, dtype: int64

********************


М    5457
Ж    4543
Name: gender, dtype: int64

********************


37    478
38    477
35    474
36    456
34    447
     ... 
92      2
88      1
82      1
85      1
83      1
Name: age, Length: 70, dtype: int64

********************


2     1048
1     1035
7     1028
8     1025
5     1012
3     1009
4      989
9      984
6      967
10     490
0      413
Name: objects, dtype: int64

********************


105473.74    2
130170.82    2
97819.36     1
120782.70    1
119703.10    1
            ..
117199.80    1
183555.24    1
137648.41    1
112689.95    1
99141.86     1
Name: balance, Length: 6381, dtype: int64

********************


1    5084
2    4590
3     266
4      60
Name: products, dtype: int64

********************


1    5151
0    4849
Name: loyalty, dtype: int64

********************


0    7963
1    2037
Name: churn, dtype: int64

********************


ValueError: Length mismatch: Expected axis has 12 elements, new values have 10000 elements

Датасет содержит 12 колонок с записями о пользователях услуг банка с корректными типами данных. 
В колонке `userid`  - 10 000 уникальных идетификаторов пользователя, повторений нет.   
Дубликатов нет.
Всего в датасете 10 000 строк в каждом столбце, кроме столбца Balance, где 6 383 значений и есть пропуски.
Проведем предварительную подготовку данных:
1. Исследуем пропущенные значения, все они находятся в графе "Баланс", которая содержит информацию о балансе счета. Баланс дебитовый и не предполагает отрицательных значений.
Добавим новый столбец, где сохраним группу с отсутствующим балансом, и исследуем средние значения параметров в этой группе


In [ ]:
bank_data['balance_nan'] = 1
bank_data.loc[bank_data['balance'] >= 0, 'balance_nan'] = 0
bank_data.groupby('balance_nan').mean()

Из таблицы можно сделать вывод, что несколько причин, в том числе сбой системы хранения информации, но следует отметить особенность - у клиентов с неизвестным балансом гораздо меньше отток и больше число используемых продуктов при близких показателях лояльности.  
Возможно, используемые продукты входят в обязательные при обслуживании таких карт. Так как баланс дебетовой карты остается нулевым, пока счет не пополнит сам клиент, его работодатель, партнер и т. д., то можно предположить зарплатный проект либо перевод поступающих денег автоматически на накопительный счет.  
В таком случае клиенту действительно будет сложнее перейти в другой банк, как правило компаниям выгодно реализовывать зарплатные проекты, а на пополняемых вкладах закрытие счета раньше времени может привести к нарушению условий договора и списанию процентов.  
Возможно изменение или на нули или на средние значения по городу/возрасту/полу. Я выбрала замену на нули, потому что достоверных данных о сбое в хранилище, в то время как деньги на сберегательном счете и деньги на дебетовой карте это все-таки не равноценно: последние можно тратить, получать кэшбек, а на первую группу начисляется равномерная и постоянная прибыль. В данных нет выделения важного параметра - наличия открытых сберегательных счетов или зарплатных проектов, хотя отток в таких группах точно будет меньше, если банк конкурентноспособен по отношению к остальным игрокам рынка.

In [ ]:
#заменим пропуски на нули
bank_data['balance'] = bank_data['balance'].fillna(0.0)
#удалим id пользователей, все значения уникальны
bank_data = bank_data.drop(columns=['userid'])

In [ ]:
bank_data.columns = inspect(bank_data)

2. Рассмотрим подробнее выведенные данные: 
*`score` - в среднем скоринговый балл составляет 650,5 ед. При этом минимальный балл 350 ед, а максимальный 850 ед. Медиана и среднее близки (650,5 и 652,0 ед), значит распределение нормальное;  
*`city` - клиенты из Ярославля, Ростова и Рыбинска;  
*`gender` - мужчин и женщин примерно пополам (55 и 45 % соответственно);  
*`age`  - средний возраст клиента 39 лет. Клиентом банка может быть только человек старше 18 лет, а максимальный возраст не ограничен, самому пожилому клиенту 92 года;  
*`objects`  - в собственности клиента в среднем 5 объектов. Следует отметить, что наличие собственности не является обязательным условием, поэтому есть клиенты, не имеющие собственности. Максимальное число объектов в собственности 10 ед;  
*`balance`  - счета клиентов дебетовые, поэтому минимально баланс может быть только нулевым, а максимальная сумма составляет 250 898,1 руб. В среднем, на счету клиента лежит около 76 485,9 руб;  
*`creditcard`  - кредитная карта есть в среднем у 70% клиентов банка;  
*`loyalty` - более половины клиентов лояльны и активны;  
*`estimated_salary` - средняя заработная плата (прогнозная) составляет 100 090 руб, медиана примерно на том же уровне (100193), скорее всего, выбросов нет, распределение нормальное. Максимальная зарплата - 199 992,5 руб, минимальная - 11,6. Возможно, клиент ошибся при заполнении какой-либо анкеты с информацией о зарплате или савтоматический расчет для временно безработных и получающих пособие по безработице клиентов показал данный результат.  
*`churn` - около 20% отток клиентов, 80% остались клиентами банка.
Для признаков возраст, число объектов в собственности, баланс счета, наличие кредитной карты, лояльность, заработная плата, отток выборка более вариабельна, чем для кредитной оценке. Это логично, так как все клиенты составляют однородную группу, пройдя оценку как потенциальные потребители банковских услуг


In [ ]:
display(bank_data.groupby('churn').agg('mean').round(2))

Можно отметить, что в среднем скоринговый балл у оставшихся клиентов выше, они моложе, число объектов в собственности у двух групп примерно одинаковое, но на балансе у удержанных клиентов средств меньше и они немного чаще пользуются продуктами банков, их лояльность заметно выше, чем у ушедших. При этом число владевших кредитными картами примерно одинаковое, в среднем 70%, различия в заработной плате не очень значительные.

### Вывод

* >датасет содержит 12 колонок с записями о пользователях услуг банка, типы данных приведены корректно. В колонке userid - 10 000 уникальных идетификаторов пользователя, повторений нет;
* >дубликатов нет;
* >есть особенность - баланс дебетовый, при этом часть значений пропущена. Баланс может быть нулевым, так как не все продукты предполагают наличие дебетового счета.  Была реализаована замена пропущенных значений на нулевые;
* >были отмечены нетипичные значения в области средней заработной платы. Выдвинуто предположение о возможном начислении пособия по безработице (некоторые выпадающие показатели близки к МРОТ), а также ошибках при заполнении анкеты.  
Анализ данных позволил определить портрет клиента банка:
 

В отношении оттока клиентов были отмечены следущие особенности (предварительно, по средним значениям относительно группы оставшихся клиентов):
* >скоринговый балл клиентов из группы оттока ниже, они старше, реже пользуются услугами банка и их лояльность ниже. 

## Визуализация данных

Построим столбчатые гистограммы и распределения признаков для тех, кто ушёл (отток) и тех, кто остался (не попали в отток):

In [ ]:
#отделяю бинарные переменные, внесла правки, избавившись от get_values()
binary_columns = ['city', 'gender', 'сredit_сard', 'loyalty', 'churn']
columns = bank_data.drop(columns=['churn', 'city', 'gender']).columns.to_numpy().tolist()

In [ ]:
#функция для построяния диаграмм
def create_hist(df, columns):
    for i in columns:
        #Строим графики
        if i not in binary_columns:
            fig, (ax1, ax2) = plt.subplots(1,2,figsize=(14,5),gridspec_kw={'width_ratios': [2, 1]})
            min_value = min(df[i])
            max_value = max(df[i])
            bins_range = np.linspace(min_value, max_value, 10) # список границ столбиков гистограмм
            sns.distplot(bank_data.query('churn == 0')[columns], kde=False, ax=ax1, color="blue", norm_hist=True, bins = bins_range)
            sns.distplot(bank_data.query('churn == 1')[columns], kde=False, ax=ax1, color="red", norm_hist=True, bins = bins_range)
            #Зададим названия
            ax1.set_title('Распределение признака {}'.format(i))
            ax1.set_xlabel('Значение')
            ax1.set_ylabel('Частота')
            ax1.legend(labels=['Остались', 'Ушли'])
            ax1.axvline(x=df[i].median(),   
                linewidth=2,
                color='red',
                linestyle='--')
            #линия 1 квантиля
            ax1.axvline(x=df[i].quantile(.025),   
                linewidth=3, 
                color='b', 
                linestyle='--')
            #линия 2 квантиля
            ax1.axvline(x=df[i].quantile(.05),   
                linewidth=3, 
                color='b', 
                linestyle='--')
            #Линия 3 квантиля
            ax1.axvline(x=df[i].quantile(.75),
                linewidth=3, 
                color='b', 
                linestyle='--')
            
            #ящик с усами
            sns.boxplot(y=i, x='churn', data=df, ax=ax2, palette="coolwarm")
            sns.stripplot(y=i, x='churn', data=df, jitter=True, zorder=1, color='#7cc3f2', ax=ax2, size=2,alpha=0.2)

            ax2.set_title("Boxplot для признака {}".format(i))
            ax2.set_xlabel('Отток')
            ax2.set_ylabel('Признак {}'.format(i))
            ax2.legend(labels=['Остались', 'Ушли'])
        
        else:
            fig, ax = plt.subplots(2, 2, figsize=(20, 15))
            for variable, subplot in zip(binary_columns, ax.flatten()):
                sns.countplot(data=bank_data, x=variable, hue='churn', ax=subplot)
                subplot.set_ylabel('Количество')
                subplot.set_xlabel('Признак {}'.format(variable))
                subplot.legend(labels=['Остались', 'Ушли'])
                for label in subplot.get_xticklabels():
                    label.set_rotation(45)   

        plt.grid()
        plt.show()  

In [ ]:
create_hist(bank_data, columns)

* >датасет содержит 12 колонок с записями о пользователях услуг банка, типы данных приведены корректно. В колонке userid - 10 000 уникальных идетификаторов пользователя, повторений нет;
* >дубликатов нет;
* >есть особенность - баланс дебетовый, при этом часть значений пропущена. Баланс может быть нулевым, так как не все продукты предполагают наличие дебетового счета.  Была реализаована замена пропущенных значений на нулевые;
* >были отмечены нетипичные значения в области средней заработной платы. Выдвинуто предположение о возможном начислении пособия по безработице (некоторые выпадающие показатели близки к МРОТ), а также ошибках при заполнении анкеты.  

Согласно представленным графическим данным:
* >баллы кредитного скоринга, возраст и баланс имеют распределение, близкое к нормальному;
* >распределение данных по балансу искажено внесенными изменениями (нулями)

Анализ данных позволил определить портрет типового клиента банка и особенности в группах оттока и удержания:
* >скоринговый балл потребителя услуг банка ~650, скоринговый балл клиентов из группы оттока ниже, ~645 баллам;
* >наиболее часто клиенты банка из Рыбинска. Среди клиентов из Ростова большее число выбивших;
* >распределение по половому признаку почти равнодолевое, как среди оттока, так и среди удержанных;
* >средний возраст клиента банка 39 лет, в группе оттока преобладают клиенты страше, ~45 лет;
* >средний доход клиента банка более 100 тыс руб, доход клиентов из группы оттока около 101 тыс руб;
* >среднее число потребляемых продуктов на клиента составляет около 1-2 услуги. Условно можно разделить клиентов на две группы: те, который используют одну услугу и те, которые выбирают сразу несколько;
* >клиенты банка в большей степени лояльны к нему. При этом в группе оттока ярко выражена сниженная лояльность;
* >типичный клиент банка предпочитает использовать кредитную карту. Это характерно для обеих групп.

Рассмотрим тепловую карту корреляции:

In [ ]:
#изменим качественные переменные для анализа
bank_data_corr = bank_data.copy()
bank_data_corr['city'] = bank_data_corr['city']
bank_data_corr['gender'] = bank_data_corr['gender']
bank_data_corr = pd.get_dummies(bank_data_corr, columns=['gender','city'])

#тепловая карта корреляции
fig, ax = plt.subplots(figsize=(10,10))
heatmap = sns.heatmap(bank_data_corr.corr().round(2), vmin=-1, vmax=1, annot=True, cmap='BrBG')

# Дайте тепловой карте название. Параметр pad (padding) определяет расстояние заголовка от верхней части тепловой карты. 
heatmap.set_title('Корреляция', fontdict={'fontsize':12}, pad=12);

In [ ]:
plt.figure(figsize=(8, 12))
heatmap = sns.heatmap(bank_data_corr.corr()[['churn']].sort_values(by='churn', ascending=False), vmin=-1, vmax=1, annot=True, cmap='BrBG')

heatmap.set_title('Особенности Корреляция с оттоком', fontdict={'fontsize':18}, pad=16);

* >заметная отрицательная корреляция между признаками лояльности и числа используемых продуктов. Чем ни больше клиент проявляется активность, а его лояльность выше и чем ни чаще он покупает дополнительные продукты, тем меньше вероятность оттока. Также меньше вероятность оттока у клиентов из Рыбинска, Ярославля. 
* >положительная корреляция наблюдается между показателями возраста, городом и балансом - чем ни старше клиент и чем ни меньше средств на его балансе, тем больше вероятность оттока. Аналогично увеличивается вероятность оттока, если клиент из Ростова Великого.
* > о влиянии показателя дохода можно предварительно сказать, что прямой и сильной взаимосвязи между показателями дохода и отока нет.  

### Вывод

* >существует тесная взаимосвязь между оттоком и признаками лояльности и числом используемых продуктов. Чем ни больше клиент проявляется активность, а его лояльность выше и чем ни чаще он покупает дополнительные продукты, тем меньше вероятность оттока.
Кроме того, положительная корреляция наблюдается между показателями возраста и баланса - чем ни старше клиент и чем ни меньше средств на его балансе, тем больше вероятность оттока.
* > следует обратить внимание на отток клиентов с нулевым балансом из Ростова Великого;
* > предварительный портрет типового клиента показал, что наиболее активными клиентами банка являются мужчины из Ярославля, с высокой лояльностью, использующие дополнительные продукты, с ненулевым балансом.

## Кластеризация клиентов

Для кластеризации выполним ряд действий:

1. Стандартизируем данные.
2. Построим матрицу расстояний функцией linkage() на стандартизованной матрице признаков и нарисуем дендрограмму.
3. Обучим модель кластеризации на основании алгоритма K-Means и спрогнозируем кластеры клиентов. 

In [ ]:
bank_data_cluster = bank_data_corr.copy()

In [ ]:
from sklearn.cluster import KMeans
from scipy.cluster.hierarchy import dendrogram, linkage 

# обязательная стандартизация данных перед работой с алгоритмами
sc = StandardScaler()
X_sc = sc.fit_transform(bank_data_cluster)

linked = linkage(X_sc, method = 'ward') 

plt.figure(figsize=(15, 10))  
dendrogram(linked, orientation='top')
plt.title('Hierarchial clustering for banki')
plt.show() 

Предлагается разбить данные на 4 более равномерных кластера.

In [ ]:
km = KMeans(n_clusters=4, random_state=0) # задаём число кластеров, равное 4, и фиксируем значение random_state для воспроизводимости результата
labels = km.fit_predict(X_sc) # применяем алгоритм к данным и формируем вектор кластеров 

Посмотрим на средние значения признаков:

In [ ]:
# сохраняем метки кластера в поле нашего датасета
bank_data_cluster['cluster_km'] = labels

# выводим статистику по средним значениям наших признаков по кластеру
clusters = bank_data_cluster.groupby('cluster_km').mean().round(3)
clusters 

 Основные закономерности в кластерах (средние):
* > кластеры соответствуют городам: кластер 3 с наибольшим оттоком распределен в Ростове Великом, где предварительно уже была установлена наибольшая активность оттока, а кластеры с меньшим оттоком, 0-2, в Ярославле и Рыбинске.
* >кластер 0. Скоринговый балл 651,3, возраст около 39 лет, в собственности около 5 объектов, на балансе более 61 тыс руб, лояльность выше среднего, число используемых продуктов 1,54 шт, кредитные карты есть у 69,5% клиентов, лояльны 53% клиентов, заработная плата более 99 тыс руб, мужчин 56%, отток 16,7%, город Рыбинск;
* >кластер 1. Клиенты кластера из г. Ярославль, скоринговый балл меньше, чем у 0 кластера, (649 баллов), эта группа клиентов немного младше (38 лет), на балансе более 60 тыс руб. В этом кластере потребляется большее число дополнительных продуктов лояльность 51,4%, отток 20,3%, все клиенты указанного сегмента - женщины;
* >кластер 2. Потребители этого сегмента из Ярославля, все мужчины. Как видим, отток указанного кластера меньше, чеи у близкого к нему кластера 1 с другим гендером и составляет всего 12,7%. Объем потребляемых услуг заметно ниже, но при этом лояльность чуть выше, чем у кластера 1;
* >кластер 3. Отток этого кластера максимален, 32,4%. Все клиенты сегмента из Ростова Великого, гендерное соотношение почти равное, лояльность заметно ниже, чем в других группах (49,7%).  
Построим распределение признаков для кластеров:

In [ ]:
import itertools

# определим функцию отрисовки графиков попарных признаков для кластеров
def show_clusters_on_plot(df, x_name, y_name, cluster_name):
    plt.figure(figsize=(5, 5))
    sns.scatterplot(
        df[x_name], df[y_name], hue=df[cluster_name], palette='Paired'
    )
    plt.ylabel('Частота')
    plt.title('{} vs {}'.format(x_name, y_name))
    plt.show()

# отрисуем графики для попарных сочетаний признаков с разметкой по кластерам
col_pairs = list(itertools.combinations(bank_data_cluster.drop('cluster_km', axis=1).columns, 2))
for pair in col_pairs:
    show_clusters_on_plot(bank_data_cluster, pair[0], pair[1], 'cluster_km')

Построим модель прогнозирования отока:

In [ ]:
# разделим наши данные на признаки (матрица X) и целевую переменную (y)
X = bank_data_cluster.drop(columns=['churn'])
y = bank_data_cluster['churn']
stratify = y

# разделяем модель на обучающую и валидационную выборку
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Обучим модель на train-выборке двумя способами:

* логистической регрессией,
* случайным лесом.  
Оценим метрики accuracy, precision и recall для обеих моделей на валидационной выборке. Сравним по ним модели.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score, precision_score, recall_score, roc_auc_score

In [ ]:
# список моделей
models = [
    LogisticRegression(random_state=0),
    RandomForestClassifier(random_state=0)
]

# функция, которая принимает на вход модель и данные и выводит метрики
def make_prediction(m, X_train, y_train, X_test, y_test):
    model = m
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(
        'Accuracy:{:.2f} Precision:{:.2f} Recall:{:.2f}'.format(
            accuracy_score(y_test, y_pred),
            precision_score(y_test, y_pred),
            recall_score(y_test, y_pred),
            roc_auc_score(y_test, y_pred)
        )
    )

In [ ]:
# цикл, который выводит метрики по списку моделей
for i in models:
    print(i)
    make_prediction(m=i, X_train=X_train, y_train=y_train,
                    X_test=X_test, y_test=y_test)

Метрика accuracy, которая описывает общую точность предсказания модели по всем классам, для случайного леса выше, как и точность и полнота. Accuracy высокая может быть обусловлена неравномерностью распределения данных об ушедших клиентах.Метрика recall более важна: может быть спрогнозировано наступление события "пользователь останется", но произойдет отток (события становятся практически равновероятностыми). Несмотря на то, что и эта метрика в модели случайного леса намного выше, этого недостаточно для практического применения модели. Возможно повышение точности модели при использовании более качественных данных.

### Вывод

Рекомендуется обратить внимание на два кластера:

* >  кластер 1. Отмечено, что клиенты указанного сегмента только женщины, они немного младше (38 лет), отток 20,3%, возможно, следует оптимизировать предложения для женщин. Например, выпустить дебетовые карты с кэшбеком за покупки в магазинах, их, как правило, чаще совершают женщины. Обратить внимание на пенсионные предложения, не смотря на молодой средний возраст группы - средняя продолжительность жизни в гендере выше и женщины чаще склонны задумываться о прибыльности пенсионных накоплений;

* > кластер 3. Отток этого кластера максимален, 32,4%. Все влиенты сегмента из Ростова Валикого, предлагается организовать региональные акции для повышения лояльности кластера.

### Проверка гипотез

Гипотеза различия среднего дохода в экспериментальных групп клиентов банка по числу используемых услуг:

* Н0 - различия среднего дохода между экспериментальными группами клиентов банка которые пользуются двумя и более продуктами банка, и теми, которые пользуются одним, статистически не значимы
* Н1 - различия среднего дохода между экспериментальными группами клиентов банка которые пользуются двумя и более продуктами банка, и теми, которые пользуются одним, статистически, статистически значимы 
Гипотеза относительно представленных данных.

In [ ]:
#рассмотрим распределения
fig, (ax1, ax2) = plt.subplots(1,2,figsize=(20,5))

sns.distplot(bank_data_cluster.query('products == 1')['estimated_salary'], kde=False, ax=ax1)
ax1.set_title("Распределение средних доходов клиентов c 1 продуктом")
ax1.set_ylabel('Частота')
ax1.set_xlabel('Доход')

sns.distplot(bank_data_cluster.query('products >= 2')['estimated_salary'], kde=False, ax=ax2)
ax2.set_title("Распределение средних доходов клиентов с 2 и более продуктами")
ax2.set_ylabel('Частота')
ax2.set_xlabel('Доход')

plt.show()

In [ ]:
alpha = .05

from statsmodels.stats import weightstats as stests

results = st.ttest_ind(
    bank_data_cluster.query('products == 1')['estimated_salary'],
    bank_data_cluster.query('products > 1')['estimated_salary'],
    equal_var=False)

print('p-значение:', results.pvalue.round(2))

if (results.pvalue < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")

Дополнительно применим метод Манна-Уитни, проверив схожесть распределений.  
Для нулевой гипотезы H0:  
`Доходы клиентов, которые пользуются двумя и более  продуктами банка, и теми, которые пользуются одним, равны`
Исходя из Н0 формулируется альтернативная гипотеза Н1:  
`Доходы клиентов, которые пользуются двумя и более продуктами банка, и теми, которые пользуются одним, различаются`

In [ ]:
alpha = 0.05

results = st.mannwhitneyu(
    bank_data.query('products == 1')['estimated_salary'], 
    bank_data.query('products == 2')['estimated_salary'], 
    alternative = 'two-sided')

print('p-значение:', results.pvalue)

if (results.pvalue < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")

Вывод: различия среднего дохода между экспериментальными группами клиентов банка использующих два и более продуктов статистически не значимы.  
Проведем исследование взаимосвязи между оттоком клиентов и средним числом потребляемых услуг.
Гипотеза различия дохода между экспериментальными группами клиентов банка по числу используемых услуг:
Н0 - различия по среднему числу потребляемых продуктов, совершившими указанное событие (уход из банка, отток), статистически не значимы  
Н1 - различия по среднему числу потребляемых продуктов, совершившими указанное событие (уход из банка, отток), статистически значимы  

In [ ]:
#проверим распределения
fig, (ax1, ax2) = plt.subplots(1,2,figsize=(20,5))

sns.distplot(bank_data_cluster.query('products == 1')['churn'], kde=False, ax=ax1)
ax1.set_title("Распределение оттока клиентов в среднем c 1 продуктом")
ax1.set_ylabel('Частота')
ax1.set_xlabel('Доход')

sns.distplot(bank_data_cluster.query('products >= 2')['churn'], kde=False, ax=ax2)
ax2.set_title("Распределение оттока клиентов в среднем с 2 и более продуктами")
ax2.set_ylabel('Частота')
ax2.set_xlabel('Доход')

plt.show()

In [ ]:
alpha = .05

results, pvalue = stests.ztest(
    bank_data.query('products == 1')['churn'], 
    x2=bank_data.query('products > 1')['churn'],
    value = 0,
    alternative='two-sided')

print('p-значение:', float(pvalue))

if (pvalue < alpha):
    print("Отвергаем нулевую гипотезу")
else:
    print("Не получилось отвергнуть нулевую гипотезу")

Отвергаем нулевую гипотезу: отток клиентов, которые в среднему пользуются двумя и более продуктами банка, и отток тех, которые пользуются одним, различаются.

## Выводы и рекомендации

1. Выводы и рекомендации по группам клиентов с наибольшим оттоком:
* основными факторами, влияющими на отток являются: географический (город), гендерный, лояльность, использование дополнительных услуг.  
В разрезе кластеров, наиболее подверженных оттоку можно дать следующие рекомендации:
* > для всех кластеров: мероприятия по повышению лояльности и числа используемых услуг.  
*`система бонусов` - начисление бонусов за оплату картой, либо, если система уже активна, ее изменение. Также можно связать систему бонусов с программой лояльнсоти. В том числе, начислять бонусы за активность клиентов в социальных сетях в группах банка, формирование положительных отзывов на специализированных сайтах, банковских маркетплейсах (https://www.banki.ru/, например);
*`формирование узнаваемого, близкого целевым сегментам имиджа` - большинству клиентов в возрастном сегменте клиентов банка наиболее близки семейные ценности и стабильность. Банк должен ассоциироваться с основными предпочтениями клиентов, его дизайн, слоган, рекламный контент соответствовать им;
*`адресность предложений` -  на основе имеющихся данных о клиентских кластеров возможна отправка более индивидуальных предложений, направленных на конкретный целевой сегмент: например, региональные акции, кредитные предложения для кластеров с более низкими заработными платами в разрезе потребительских краткосрочных займов, предложения о долгосрочных вкладах для кластеров с большими суммами на счету и пр.
* `повышение количества продаваемых продуктов и услуг на одного клиента` - выявлено, что с числом услуг растет и вероятность закрепления клиента. Рекомендуемыми акциями могут быть: начисление повышенных бонусов за использование дополнительных услуг, перекрестные продажи взаимовыгодных продуктов (вместе с услугой вклада с возможностью пополнения, например, услуга копилка, которая бы переводила закрепленную сумму ежемесячно);
* > для отдельных кластеров:
* `кластер 1` - основной фактор оттока гендерный. Отмечено, что клиенты указанного сегмента только женщины, они немного младше (38 лет), отток 20,3%, возможно, следует оптимизировать предложения для женщин.  
Примерные мероприятия для удержания кластера:
* выпустить дебетовые карты с кэшбеком за покупки в сетевых магазинах, их, как правило, чаще совершают женщины;
* сформировать предложения услуг для женщин в декретном отпуске (кредитные предложения, дебетовые услуги обслуживания);
* индивидуальные пенсионные предложения, не смотря на молодой средний возраст группы - средняя продолжительность жизни в гендере выше и женщины чаще склонны задумываться о прибыльности пенсионных накоплений.
* `кластер 3` - отток этого кластера максимален, 32,4%. Все клиенты сегмента из Ростова Великого, предлагается организовать региональные акции для повышения лояльности кластера. Особое внимание требуется обратить на причины оттока и првести проверку качества обслуживания.

## Дашборд и презентация

Ссылка на дашборд: https://public.tableau.com/app/profile/mariannmois/viz/_16535738693460/Dashboard1   
Ссылка на презентацию: https://disk.yandex.ru/i/QXuJJIBhTjIpeQ